### PR TITLE
Use DCR attributes in the token and authorization flows

### PR DESCRIPTION
**Pull request by SachiniSiriwardene**
### Proposed changes in this pull request

[List all changes you want to add here. If you fixed an issue, please
add a reference to that issue as well.]

-

### When should this PR be merged

[Please describe any preconditions that need to be addressed before we
can merge this pull request.]

### Follow up actions

[List any possible follow-up actions here; for instance, testing data
migrations, software that we need to install on staging and production
environments.]

-

#### General

- [ ] **Is this PR explained thoroughly?** All code changes must be accounted for in the PR description.
- [ ] **Is the PR labeled correctly?**

#### Functionality

- [ ] **Are all requirements met?** Compare implemented functionality with the requirements specification.
- [ ] **Does the UI work as expected?** There should be no Javascript errors in the console; all resources should load. There should be no unexpected errors. Deliberately try to break the feature to find out if there are corner cases that are not handled.

#### Code

- [ ] **Do you fully understand the introduced changes to the code?** If not ask for clarification, it might uncover ways to solve a problem in a more elegant and efficient way.
- [ ] **Does the PR introduce any inefficient database requests?** Use the debug server to check for duplicate requests.
- [ ] **Are all necessary strings marked for translation?** All strings that are exposed to users via the UI must be [marked for translation](https://docs.djangoproject.com/en/1.10/topics/i18n/translation/).

#### Tests

- [ ] **Are there sufficient test cases?** Ensure that all components are tested individually; models, forms, and serializers should be tested in isolation even if a test for a view covers these components.
- [ ] **If this is a bug fix, are tests for the issue in place?**  There must be a test case for the bug to ensure the issue won’t regress. Make sure that the tests break without the new code to fix the issue.
- [ ] **If this is a new feature or a significant change to an existing feature?** has the manual testing spreadsheet been updated with instructions for manual testing?

#### Security

- [ ] **Confirm this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.**
- [ ] **Are all UI and API inputs run through forms or serializers?**
- [ ] **Are all external inputs validated and sanitized appropriately?**
- [ ] **Does all branching logic have a default case?**
- [ ] **Does this solution handle outliers and edge cases gracefully?**
- [ ] **Are all external communications secured and restricted to SSL?**

#### Documentation

- [ ] **Are changes to the UI documented in the platform docs?** If this PR introduces new platform site functionality or changes existing ones, the changes should be documented.
- [ ] **Are changes to the API documented in the API docs?** If this PR introduces new API functionality or changes existing ones, the changes must be documented.
- [ ] **Are reusable components documented?** If this PR introduces components that are relevant to other developers (for instance a mixin for a view or a generic form) they should be documented in the Wiki.

### CODE REVIEW

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java [hunk 1]
```diff
@@ -2514,6 +2517,11 @@ private void handleOIDCRequestObject(OAuthMessage oAuthMessage, OAuthAuthzReques
 
         validateRequestObjectParams(oauthRequest);
         String requestObjValue = null;
+        // Check whether the application mandates PAR request.
+        if (isPARMandated(parameters.getClientId(), parameters.getTenantDomain()) && !isRequestUri(oauthRequest)) {
```
Comment (janakamarasena) [inline components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java:2521]: We build the par request over here https://github.com/wso2-extensions/identity-inbound-auth-oauth/blob/master/components/org.wso2.carbon.identity.oauth.par/src/main/java/org/wso2/carbon/identity/oauth/par/core/ParRequestBuilder.java#L69

If a request is a par request then this canHandler will return true https://github.com/wso2-extensions/identity-inbound-auth-oauth/blob/master/components/org.wso2.carbon.identity.oauth.par/src/main/java/org/wso2/carbon/identity/oauth/par/core/ParRequestBuilder.java#L67

Lets check on the possibility of moving the par mandatory check to https://github.com/wso2-extensions/identity-inbound-auth-oauth/blob/master/components/org.wso2.carbon.identity.oauth.par/src/main/java/org/wso2/carbon/identity/oauth/par/core/ParRequestBuilder.java#L69

One way is to check if par is mandatory check whether the Par request builder has been selected by the can handler

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/responsemode/provider/jarm/JarmResponseModeProvider.java [hunk 2]
```diff
@@ -162,8 +162,9 @@ protected String getJWTToken(AuthorizationResponseDTO authorizationResponseDTO,
 
     protected static JWSAlgorithm getJWTSignatureAlgorithm() throws OAuthSystemException {
 
+        String sigAlg;
         JWSAlgorithm signatureAlgorithm = new JWSAlgorithm(JWSAlgorithm.NONE.getName());
-        String sigAlg = OAuthServerConfiguration.getInstance().getIdTokenSignatureAlgorithm();
+        sigAlg = OAuthServerConfiguration.getInstance().getIdTokenSignatureAlgorithm();
```
Comment (janakamarasena) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/responsemode/provider/jarm/JarmResponseModeProvider.java:167]: Lets revert this change.

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/openidconnect/DefaultIDTokenBuilder.java [hunk 3]
```diff
@@ -324,6 +324,10 @@ private String getIDToken(String clientId, String spTenantDomain, JWTClaimsSet j
                     encryptionAlgorithm, encryptionMethod, spTenantDomain,
                     clientId).serialize();
         } else {
+            if (StringUtils.isNotEmpty(oAuthAppDO.getIdTokenSignatureAlgorithm())) {
+                signatureAlgorithm = OAuth2Util.mapSignatureAlgorithmForJWSAlgorithm(
+                        oAuthAppDO.getIdTokenSignatureAlgorithm());
+            }
```
Comment (janakamarasena) [inline components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/openidconnect/DefaultIDTokenBuilder.java:330]: Should not alter the class level signatureAlgorithm variable. If the DefaultIDTokenBuilder object is reused erroneous results will be produced.

Also isn't this resolution of the signature alg applicable for all the places? Better to modify the methods to pass in the sig alg from the entry points

Comment (SachiniSiriwardene) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/openidconnect/DefaultIDTokenBuilder.java]: assigned to a method local variable - https://github.com/wso2-extensions/identity-inbound-auth-oauth/pull/2197/commits/b9eae93266ce28d2891e7f5372b798fb5634fb5c#diff-0441dee1c46a8372bf791b08b2629354ad01d0dd82d9dfbcb36b0472bfb7b752R320

File: components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/responsemode/provider/jarm/JarmResponseModeProvider.java (hunk 2 above)
Comment (SachiniSiriwardene) [reply components/org.wso2.carbon.identity.oauth/src/main/java/org/wso2/carbon/identity/oauth2/responsemode/provider/jarm/JarmResponseModeProvider.java]: removed

File: components/org.wso2.carbon.identity.oauth.par/src/main/java/org/wso2/carbon/identity/oauth/par/core/ParRequestBuilder.java [hunk 4]
```diff
@@ -83,6 +83,9 @@ public boolean canHandle(HttpServletRequest request) {
                     .logDetailLevel(DiagnosticLog.LogDetailLevel.INTERNAL_SYSTEM);
             LoggerUtils.triggerDiagnosticLogEvent(diagnosticLogBuilder);
         }
+        if (canHandle) {
+            request.setAttribute(OAuthConstants.IS_PUSH_AUTHORIZATION_REQUEST, true);
```
Comment (janakamarasena) [inline components/org.wso2.carbon.identity.oauth.par/src/main/java/org/wso2/carbon/identity/oauth/par/core/ParRequestBuilder.java:87]: Lets add this to the buildRequest method so we know the request was build from it

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java [hunk 5]
```diff
@@ -2045,6 +2045,20 @@ private String handleOAuthAuthorizationRequest(OAuthMessage oAuthMessage)
...
+        } catch (IdentityOAuth2Exception | InvalidOAuthClientException e) {
+            throw new InvalidRequestException(e.getMessage(), e.getErrorCode());
+        }
+        if (oAuthAppDO.isRequirePushedAuthorizationRequests()) {
+            if (!Boolean.parseBoolean(oauthRequest.getParam(OAuthConstants.IS_PUSH_AUTHORIZATION_REQUEST))) {
+                throw new InvalidRequestException("PAR request is mandatory for the application.",
+                        OAuth2ErrorCodes.INVALID_REQUEST,
+                        OAuth2ErrorCodes.OAuth2SubErrorCodes.INVALID_AUTHORIZATION_REQUEST);
+            }
+        }
```
Comment (janakamarasena) [inline components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java:2061]: Lets move this validation part to a private method

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java [hunk 6]
```diff
@@ -2045,6 +2045,20 @@ private String handleOAuthAuthorizationRequest(OAuthMessage oAuthMessage)
...
         }
+        // Check whether PAR should be mandated in  the request.
+        OAuthAppDO oAuthAppDO;
+        try {
+            oAuthAppDO = OAuth2Util.getAppInformationByClientId(params.getClientId(), params.getTenantDomain());
+        } catch (IdentityOAuth2Exception | InvalidOAuthClientException e) {
+            throw new InvalidRequestException(e.getMessage(), e.getErrorCode());
+        }
+        if (oAuthAppDO.isRequirePushedAuthorizationRequests()) {
+            if (!Boolean.parseBoolean(oauthRequest.getParam(OAuthConstants.IS_PUSH_AUTHORIZATION_REQUEST))) {
```
Comment (janakamarasena) [inline components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java:2056]: Lets get this directly from the request
```suggestion
            if (Boolean.TRUE.equals(oAuthMessage.getRequest().getAttribute(OAuthConstants.IS_PUSH_AUTHORIZATION_REQUEST))) {
```

Comment (SachiniSiriwardene) [reply components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java]: fixed

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java (hunk 5 above)
Comment (SachiniSiriwardene) [reply components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java]: moved

File: components/org.wso2.carbon.identity.oauth.par/src/main/java/org/wso2/carbon/identity/oauth/par/core/ParRequestBuilder.java (hunk 4 above)
Comment (SachiniSiriwardene) [reply components/org.wso2.carbon.identity.oauth.par/src/main/java/org/wso2/carbon/identity/oauth/par/core/ParRequestBuilder.java]: changed

**Review by janakamarasena — APPROVED** [review]
